From 9564dbb645a4be08aff7b65ac35b56e185a7598c Mon Sep 17 00:00:00 2001
From: Jordan Williams <jordan@jwillikers.com>
Date: Tue, 7 Nov 2023 11:00:39 -0600
Subject: [PATCH] Use find_library to locate GL libraries not provided by
 FindOpenGL

FindOpenGL is used to find the OpenGL libraries.
It doesn't find all of the necessary libraries, however.
These missing libraries have been added to the link line directly.
This is problematic when FindOpenGL is pointed at OpenGL libraries somewhere outside LD_LIBRARY_PATH and the system's default linker paths.
In such cases, the libraries won't be found at link time.

CMake's find_library function can be used instead to add the libraries as this commit does.
This function can use the location of the GL libraries that were already found as a hint.
Furthermore, this allows consumers to use CMake cache variables to set the exact libraries to use if needed.
---
 CMakeLists.txt | 21 +++++++++++++++++----
 1 file changed, 17 insertions(+), 4 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index a2a95c02..772e73f3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -263,13 +263,23 @@ IF(FREEGLUT_GLES)
     LIST(APPEND PUBLIC_DEFINITIONS -DFREEGLUT_GLES)
     if(NOT CMAKE_VERSION VERSION_LESS "3.27")
       FIND_PACKAGE(OpenGL REQUIRED COMPONENTS EGL GLES2 OpenGL)
-      LIST(APPEND LIBS GLESv1_CM OpenGL::EGL OpenGL::GLES2 OpenGL::OpenGL)
+      CMAKE_PATH(GET OPENGL_gles2_LIBRARY PARENT_PATH _OPENGL_LIBDIR)
+      FIND_LIBRARY(GLES1_LIBRARY GLESv1_CM HINTS ${_OPENGL_LIBDIR} REQUIRED)
+      LIST(APPEND LIBS ${GLES1_LIBRARY} OpenGL::EGL OpenGL::GLES2 OpenGL::OpenGL)
     elseif(NOT CMAKE_VERSION VERSION_LESS "3.10")
       FIND_PACKAGE(OpenGL REQUIRED COMPONENTS EGL OpenGL)
-      LIST(APPEND LIBS GLESv2 GLESv1_CM OpenGL::EGL OpenGL::OpenGL)
+      GET_FILENAME_COMPONENT(_OPENGL_LIBDIR ${OPENGL_egl_LIBRARY} DIRECTORY)
+      FIND_LIBRARY(GLES1_LIBRARY GLESv1_CM HINTS ${_OPENGL_LIBDIR})
+      FIND_LIBRARY(GLES2_LIBRARY GLESv2 HINTS ${_OPENGL_LIBDIR})
+      LIST(APPEND LIBS ${GLES1_LIBRARY} ${GLES2_LIBRARY} OpenGL::EGL OpenGL::OpenGL)
     else()
       FIND_PACKAGE(OpenGL REQUIRED)
-      LIST(APPEND LIBS EGL GLESv2 GLESv1_CM)
+      LIST(GET ${OPENGL_LIBRARIES} 0 _OPENGL_LIB)
+      GET_FILENAME_COMPONENT(_OPENGL_LIBDIR ${_OPENGL_LIB} DIRECTORY)
+      FIND_LIBRARY(EGL_LIBRARY EGL HINTS ${_OPENGL_LIBDIR})
+      FIND_LIBRARY(GLES1_LIBRARY GLESv1_CM HINTS ${_OPENGL_LIBDIR})
+      FIND_LIBRARY(GLES2_LIBRARY GLESv2 HINTS ${_OPENGL_LIBDIR})
+      LIST(APPEND LIBS ${EGL_LIBRARY} ${GLES1_LIBRARY} ${GLES2_LIBRARY})
     endif()
 ELSE()
     if(NOT CMAKE_VERSION VERSION_LESS "3.10")
@@ -291,7 +301,10 @@ IF(FREEGLUT_WAYLAND)
     LIST(APPEND LIBS OpenGL::EGL OpenGL::OpenGL)
   else()
     FIND_PACKAGE(OpenGL REQUIRED)
-    LIST(APPEND LIBS EGL)
+    LIST(GET ${OPENGL_LIBRARIES} 0 _OPENGL_LIB)
+    GET_FILENAME_COMPONENT(_OPENGL_LIBDIR ${_OPENGL_LIB} DIRECTORY)
+    FIND_LIBRARY(EGL_LIBRARY EGL HINTS ${_OPENGL_LIBDIR})
+    LIST(APPEND LIBS ${EGL_LIBRARY})
   endif()
   if(NOT CMAKE_VERSION VERSION_LESS "3.6")
     PKG_CHECK_MODULES(wayland-client REQUIRED IMPORTED_TARGET wayland-client)
-- 
2.41.0

